File: MatchSticks.java
//Time - O(n^n)
//space - O(n)
class Solution {
    public boolean makesquare(int[] matchsticks) {
        int sum =0;
        for(int i=0; i< matchsticks.length; i++){
            sum += matchsticks[i];
        }
        Arrays.sort(matchsticks);
        reverse(matchsticks, matchsticks.length);
        if(sum %4 !=0) return false;
        return backtrack(matchsticks, 0, new int[4], sum/4);
    }
    boolean backtrack(int[] matchsticks, int i, int[] sub, int size){

        if(sub[0] ==size && sub[1] ==size && sub[2] ==size && sub[3] ==size) return true;
        //action
        for(int j=0; j<sub.length;j++){
            if(sub[j]+matchsticks[i] <= size){
                sub[j]+=matchsticks[i];
                if(backtrack(matchsticks,i+1 ,sub,size)) return true;;
                sub[j]-=matchsticks[i];
            }
        }
        return false;
    }

     void reverse(int a[], int n)
    {
        int left =0;
        int right = a.length-1;
        while(left<right){
            int temp = a[left];
            a[left] = a[right];
            a[right] = temp;
            ++left;
            --right;
        }
  
    }
}

File: matchsticks.py
# Time complexity: O(n)
# Space complexity: O(n)-recursive stack space
# Approach: create a array of size 4 and keep adding the value at index in array till its not larger than the maxlength of a side
# Do for loop based backtracking.


class Solution:
    def makesquare(self, matchsticks: List[int]) -> bool:
        sums = sum(matchsticks)
        if sums < 4:
            return False
        if sums % 4 != 0:
            return False
        side = sums//4
        array = [0 for i in range(4)]
        matchsticks.sort(reverse = True)
        return self.backtrack(matchsticks,0,side,array)
       
    
    def backtrack(self, matchsticks, index,side,array):
        if(index == len(matchsticks)):
            if(array[0] == side and array[1] == side and array[2]== side and array[3]==side):
                return True
            return False
    
        for i in range(4):
            if(array[i]+matchsticks[index] <= side):
                array[i] += matchsticks[index]
                if(self.backtrack(matchsticks, index+1, side, array)):
                    return True
                array[i] -= matchsticks[index]
        return False
    
    
    
    